fnlb/lb/allgrouper.go

```diff
@@ -117,69 +104,87 @@ func (a *allGrouper) remove(ded string) error {
 
 func (a *allGrouper) publishHealth() {
 
-	// get a list of healthy nodes
 	a.nodeLock.Lock()
-	new_list := make([]string, 0, len(a.nodeList))
+
+	// get a list of healthy nodes
+	newList := make([]string, 0, len(a.nodeList))
 	for key, value := range a.nodeList {
 		if value.healthy {
-			new_list = append(new_list, key)
+			newList = append(newList, key)
 		}
 	}
-	a.nodeLock.Unlock()
 
-	// sort and set the healty list pointer
-	sort.Strings(new_list)
-	a.healthyList.Store(&new_list)
+	// sort and update healthy List
+	sort.Strings(newList)
+	a.nodeHealthyList = newList
+
+	a.nodeLock.Unlock()
 }
 
 // return a copy
 func (a *allGrouper) List(string) ([]string, error) {
 
-	// safe atomic load on the current healthy list
-	ptr := a.healthyList.Load().(*[]string)
-	if len(*ptr) == 0 {
-		return nil, ErrNoNodes
-	}
+	a.nodeLock.RLock()
+	ret := make([]string, len(a.nodeHealthyList))
+	copy(ret, a.nodeHealthyList)
+	a.nodeLock.RUnlock()
 
-	ret := make([]string, len(*ptr))
-	copy(ret, *ptr)
-	return ret, nil
+	var err error
+	if len(ret) == 0 {
+		err = ErrNoNodes
+	}
+	return ret, err
 }
 
-func (a *allGrouper) runHealthCheck(iteration *uint64) {
+func (a *allGrouper) runHealthCheck() {
 
 	// fetch a list of nodes from DB
 	list, err := a.db.List()
 	if err != nil {
 		// if DB fails, the show must go on, report it but perform HC
 		logrus.WithError(err).Error("error checking db for nodes")
+
+		// compile a list of nodes to be health checked
+		a.nodeLock.RLock()
+		list = make([]string, 0, len(a.nodeList))
+		for key, _ := range a.nodeList {
+			list = append(list, key)
+		}
+		a.nodeLock.RUnlock()
+
 	} else {
-		*iteration++
+
 		isChanged := false
+
+		// compile a map of DB nodes for deletion check
+		deleteCheck := make(map[string]bool, len(list))
+		for _, node := range list {
+			deleteCheck[node] = true
+		}
+
 		a.nodeLock.Lock()
-		// handle existing & new nodes
+
+		// handle new nodes
 		for _, node := range list {
 			_, ok := a.nodeList[node]
-			if ok {
-				// mark existing node
-				a.nodeList[node].intervalID = *iteration
-			} else {
+			if !ok {
 				// add new node
-				a.nodeList[node] = &nodeState{
-					name:       node,
-					intervalID: *iteration,
-					healthy:    true,
+				a.nodeList[node] = nodeState{
+					healthy: true,
```

Comment:
we should probably start nodes in an unhealthy state until they pass a health check? prevent sending to a bad node if e.g. the wrong address got put in for a node.

Reply:
ah, i guess implicitly we're not adding to the healthy list which we have separate, so this is the case in practice even if this bit is flipped to true.

Reply:
well, after another think, it's confusing I guess. one of the pingers which changes state could end up updating the list of healthy nodes from publishHealth and include this node even before it was pinged, since this bit is true -- so it seems like we should flip it off until a check passes.

Reply:
So, this is actually something I wanted to ask, add() operation runs 1 ping, then adds it to DB. I think we can reevaluate these and how we should handle, eg. case of fnlb crash/restart, case where healthy threshold is configured higher than 1 with long interval (potentially can take minutes to become healthy), perhaps default healthy/unhealthy preference can be in DB? etc... or we could choose to always gate servers for one run of ping for all these cases to ensure quick start up time... I don't know. :-)

Reply:
do you think it's reasonable to always guard initial state with 1 ping? That gives us reasonable restart fnlb time and guards against routing requests to faulty servers. And if fnlb is configured with eg. healthy threshold of 2 and check interval of 5 secs, it nicely guards frequent state flip/flops as configured, but also ensures fast restart/start.

Reply:
that's a good point about restart. it seems like we should be able to handle these cases separately with some maneuvering (concretely, somewhat as pointed out): 

* on startup we can just load in the entire list and do one health check immediately and go? ignore healthy count / interval.
* when a user detects in `runHealthcheck` that a new node has come in, go through the normal health checks to assert health.

i have been trying to consciously avoid adding anything more than the node list to the db, if possible, but not for any good reason beyond having to manage shared state between multiple processes (not a bad reason) as well as handling partitioning. one issue with requiring a ping is that on startup, the backing list we use for consistent hashing will be modified basically N times and could lead to a small amount of thrashing -- we could add a tiny buffer in recovery time before we allow serving traffic to mitigate this. 

in any case, I do think it's a good idea to guard with 1 ping. even if we manage 'health' in a db, we could have broken links between some nodes that other load balancers may not have. I agree, I think we'll have pretty reasonable restart times doing that and we can do the proper health checking from operating state.

Reply:
sending a PR soon. (I meant: not health state in DB, but the config option of default health state for new node in DB, as config option.)

Reply:
https://github.com/fnproject/fn/pull/508